pkg/frontend/frontend_test.go

```diff
@@ -99,7 +101,139 @@ func TestFrontend_RequestHostHeaderWhenDownstreamURLIsConfigured(t *testing.T) {
+			clientCluster:      "cluster",
+			serverCluster:      "cluster",
+			enabled:            true,
+			expectedStatusCode: 200,
+		},
+		"if server and client have different cluster validation labels and soft cluster validation is enabled no error is returned": {
+			clientCluster:      "client-cluster",
+			serverCluster:      "server-cluster",
+			enabled:            true,
+			softValidation:     true,
+			expectedStatusCode: 200,
+		},
+		"if server and client have different cluster validation labels and cluster validation is enabled an error is returned": {
+			clientCluster:      "client-cluster",
+			serverCluster:      "server-cluster",
+			enabled:            true,
+			softValidation:     false,
+			expectedStatusCode: 500,
+			expectedError:      `request rejected by the server: rejected request with wrong cluster validation label "client-cluster" - it should be "server-cluster"`,
+			expectedMetrics: `
+				# HELP cortex_client_request_invalid_cluster_validation_labels_total Number of requests with invalid cluster validation label.
+        	    # TYPE cortex_client_request_invalid_cluster_validation_labels_total counter
+        	    cortex_client_request_invalid_cluster_validation_labels_total{client="query-frontend",method="/api/v1/query_range",protocol="http"} 1
+			`,
+		},
+		"if client has no cluster validation label and soft cluster validation is enabled no error is returned": {
+			clientCluster:      "",
+			serverCluster:      "server-cluster",
+			enabled:            true,
+			softValidation:     true,
+			expectedStatusCode: 200,
+		},
+		"if client has no cluster validation label and cluster validation is enabled an error is returned": {
+			clientCluster:      "",
+			serverCluster:      "server-cluster",
+			enabled:            true,
+			softValidation:     false,
+			expectedStatusCode: 511,
+			expectedError:      "rejected request with empty cluster validation label - it should be \\\"server-cluster\\\"",
+		},
+	}
+	for testName, testCase := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			// Create an HTTP server listening locally. This server mocks the downstream
+			// Prometheus API-compatible server.
+			downstreamListen, err := net.Listen("tcp", "localhost:0")
+			require.NoError(t, err)
+
+			var handler http.Handler = http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+				_, err := w.Write([]byte(responseBody))
+				require.NoError(t, err)
+			})
+			logger := log.NewNopLogger()
+			if testCase.enabled {
+				handler = middleware.ClusterValidationMiddleware(testCase.serverCluster, []string{}, testCase.softValidation, logger).Wrap(handler)
+			}
+			downstreamServer := http.Server{
+				Handler:      handler,
+				ReadTimeout:  10 * time.Second,
+				WriteTimeout: 10 * time.Second,
+			}
+
+			defer downstreamServer.Shutdown(context.Background()) //nolint:errcheck
+			go downstreamServer.Serve(downstreamListen)           //nolint:errcheck
```

Comment:
To make sure that goroutines don't hang around after tests finish (never fun to debug, if it should become a problem), please use a `WaitGroup`:
```suggestion
			var wg sync.WaitGroup
			wg.Add(1)
			go func() {
				defer wg.Done()
				downstreamServer.Serve(downstreamListen)           //nolint:errcheck
			}()
			t.Cleanup(wg.Wait)
```

Reply:
I have fixed this